File: 344. Reverse String/344.py
class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """

        """
        Title: 344. Reverse String
        Time Complexity: 
            - O(N)

        Algorithm: 
            1. Recursion

        Note:
            1. s[::-1] will copy a new list instance, 
               which is not in-place

        Date: 07/15/2021 18:44		
        Time: 192 ms  -- beat 87.99%
        Memory: 18.3 MB  -- beat 94.65%
        """

        i, j = 0, len(s)-1
        while i < j:
            s[i], s[j] = s[j], s[i]
            i += 1
            j -= 1

File: original/344. Reverse String /344.cpp
#include <vector>
using namespace std;

class Solution {
    /*
        Title: 344. Reverse String
        Time Complexity: O(N)
        Note:
            1. Two Pointer
            2. Mirroring Reverse

        Date: 12/07/2021 11:39	
        Time: 12 ms  -- beat 99.15%
        Memory: 23.3 MB  -- beat 40.05%
    */
public:
    void reverseString(vector<char>& s) {
        int left = 0, right = s.size() - 1;
        char temp;
        
        while (left < right) {
            temp = s[left];
            s[left] = s[right];
            s[right] = temp;
            
            left += 1;
            right -= 1;
        }
    }
};